block/manager.go

```diff
@@ -0,0 +1,286 @@
+package block
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/tendermint/tendermint/proxy"
+	tmtypes "github.com/tendermint/tendermint/types"
+
+	"github.com/celestiaorg/optimint/config"
+	"github.com/celestiaorg/optimint/da"
+	"github.com/celestiaorg/optimint/log"
+	"github.com/celestiaorg/optimint/mempool"
+	"github.com/celestiaorg/optimint/state"
+	"github.com/celestiaorg/optimint/store"
+	"github.com/celestiaorg/optimint/types"
+)
+
+// Manager is responsible for aggregating transactions into blocks.
+type Manager struct {
+	lastState state.State
+
+	conf    config.BlockManagerConfig
+	genesis *tmtypes.GenesisDoc
+
+	proposerKey crypto.PrivKey
+
+	store    store.Store
+	executor *state.BlockExecutor
+
+	dalc      da.DataAvailabilityLayerClient
+	retriever da.BlockRetriever
+
+	HeaderOutCh chan *types.Header
+	HeaderInCh  chan *types.Header
+
+	syncTarget uint64
+	blockInCh  chan *types.Block
+	retrieveCh chan uint64
+	syncCache  map[uint64]*types.Block
+
+	logger log.Logger
+}
+
+// getInitialState tries to load lastState from Store, and if it's not available it reads GenesisDoc.
+func getInitialState(store store.Store, genesis *tmtypes.GenesisDoc) (state.State, error) {
+	s, err := store.LoadState()
+	if err != nil {
+		s, err = state.NewFromGenesisDoc(genesis)
+	}
+	return s, err
+}
+
+func NewManager(
+	proposerKey crypto.PrivKey,
+	conf config.BlockManagerConfig,
+	genesis *tmtypes.GenesisDoc,
+	store store.Store,
+	mempool mempool.Mempool,
+	proxyApp proxy.AppConnConsensus,
+	dalc da.DataAvailabilityLayerClient,
+	logger log.Logger,
+) (*Manager, error) {
+	s, err := getInitialState(store, genesis)
+	if err != nil {
+		return nil, err
+	}
+
+	proposerAddress, err := proposerKey.GetPublic().Raw()
+	if err != nil {
+		return nil, err
+	}
+
+	exec := state.NewBlockExecutor(proposerAddress, conf.NamespaceID, mempool, proxyApp, logger)
+
+	agg := &Manager{
+		proposerKey: proposerKey,
+		conf:        conf,
+		genesis:     genesis,
+		lastState:   s,
+		store:       store,
+		executor:    exec,
+		dalc:        dalc,
+		retriever:   dalc.(da.BlockRetriever), // TODO(tzdybal): do it in more gentle way (after MVP)
+		HeaderOutCh: make(chan *types.Header),
+		HeaderInCh:  make(chan *types.Header),
+		blockInCh:   make(chan *types.Block),
+		retrieveCh:  make(chan uint64),
+		syncCache:   make(map[uint64]*types.Block),
+		logger:      logger,
+	}
+
+	return agg, nil
+}
+
+func (m *Manager) SetDALC(dalc da.DataAvailabilityLayerClient) {
+	m.dalc = dalc
+	m.retriever = dalc.(da.BlockRetriever)
+}
+
+func (m *Manager) AggregationLoop(ctx context.Context) {
+	timer := time.NewTimer(m.conf.BlockTime)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+			start := time.Now()
+			err := m.publishBlock(ctx)
+			if err != nil {
+				m.logger.Error("error while publishing block", "error", err)
+			}
+			timer.Reset(m.getRemainingSleep(start))
+		}
+	}
+}
```

Comment:
So this PR is not only about block sync as the title suggests but also contains aggregating and publishing the block to the DA layer.

Reply:
I see. Thanks for the clarification.